mmdet3d/models/roi_heads/mask_heads/primitive_head.py

```diff
@@ -0,0 +1,821 @@
+
+    def check_upright(self, para_points):
+        """Check whether is upright corrdinate.
+
+        Args:
+            para_points (torch.Tensor): Points of input.
+
+        Returns:
+            Bool: Flag of result.
+        """
+        return (para_points[0][-1] == para_points[1][-1]) and (
+            para_points[1][-1]
+            == para_points[2][-1]) and (para_points[2][-1]
+                                        == para_points[3][-1])
+
+    def check_z(self, plane_equ, para_points):
+        return torch.sum(para_points[:, 2] +
+                         plane_equ[-1]) / 4.0 < self.train_cfg['lower_thresh']
+
+    def match_point2line(self, points, xmin, xmax, ymin, ymax):
+        """Match points to corresponding edge.
+
+        Args:
+            points (torch.Tensor): Points of input.
+            xmin (float): Min of X-axis.
+            xmax (float): Max of X-axis.
+            ymin (float): Min of Y-axis.
+            ymax (float): Max of Y-axis.
+
+        Returns:
+            Tuple: Flag of matching correspondence.
+        """
+        sel1 = torch.abs(points[:, 0] - xmin) < self.train_cfg['line_thresh']
+        sel2 = torch.abs(points[:, 0] - xmax) < self.train_cfg['line_thresh']
+        sel3 = torch.abs(points[:, 1] - ymin) < self.train_cfg['line_thresh']
+        sel4 = torch.abs(points[:, 1] - ymax) < self.train_cfg['line_thresh']
+        return sel1, sel2, sel3, sel4
+
+    def compute_primitivesem_loss(self, primitive_center, primitive_semantic,
+                                  semantic_scores, num_proposal,
+                                  gt_primitive_center, gt_primitive_semantic,
+                                  gt_sem_cls_label, gt_primitive_mask):
+        """Compute loss of primitive module.
+
+        Args:
+            primitive_center (torch.Tensor): Pridictions of primitive center.
+            primitive_semantic (torch.Tensor): Pridictions of primitive
+                semantic.
+            semantic_scores (torch.Tensor): Pridictions of primitive
+                semantic scores.
+            num_proposal (int): The number of primitive proposal.
+            gt_primitive_center (torch.Tensor): Ground truth of
+                primitive center.
+            gt_votes_sem (torch.Tensor): Ground truth of primitive semantic.
+            gt_sem_cls_label (torch.Tensor): Ground truth of primitive
+                semantic class.
+            gt_primitive_mask (torch.Tensor): Ground truth of primitive mask.
+
+        Returns:
+            Tuple: Loss of primitive module.
+        """
+        batch_size = primitive_center.shape[0]
+        vote_xyz_reshape = primitive_center.view(batch_size * num_proposal, -1,
+                                                 3)
+        center_loss = self.center_loss(
+            vote_xyz_reshape,
+            gt_primitive_center,
+            dst_weight=gt_primitive_mask.view(batch_size * num_proposal, 1))[1]
+        center_loss = center_loss.sum() / (
```

Comment:
The `loss.sum()/xxx` is unnecessary, we can set `reduction='sum'` and pass `torch.sum(gt_primitive_mask.float()) + 1e-6` as avg_factor to the loss to simplify the code, so as other losses.